### ISSUE TITLE
Hyphenation disabling

### ISSUE THREAD

**Issue by wehro** [reutenauer/polyglossia#312]
The current manual states:

> polyglossia provides two functions: \disablehyphenation and \enablehyphenation . Note that when you select a new language, hyphenation will be in the same state (enabled or disabled) as before. When you reenable it, it will take the last selected language.

However, in the following document the German text *is* hyphenated.
```
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{english}
\setotherlanguage{german}

\begin{document}
This text is hyphenated.
\showhyphens{hyphenation}
\disablehyphenation
This text is not hyphenated.
\showhyphens{hyphenation}
\selectlanguage{german}
This text should not be hyphenated, but it is.
\showhyphens{Silbentrennung}
\end{document}
```
Either the package code or the manual should be corrected.

**Comment by jspitz** [comment]
Is this both with XeTeX and LuaTeX?

**Comment by wehro** [comment]
Yes, the behaviour is the same for both.

**Comment by jspitz** [comment]
Judging from the current code, hyphenation is supposed to be reset with new language, so maybe the manual is unclear.

Maintaining the state would easy:
```tex
diff --git a/tex/polyglossia.sty b/tex/polyglossia.sty
index e628aa5..2a03222 100644
--- a/tex/polyglossia.sty
+++ b/tex/polyglossia.sty
@@ -1323,6 +1323,11 @@
   \use@localhyphenmins{#1}%
   \polyglossia@lang@frenchspacing:n{#1}
   \polyglossia@lang@setdirection:n{#1}
+  % If \disablehyphenation was active, re-disable hyphenation
+  \ifbool{xpg@hyphenation@disabled}{%
+      \xdef\xpg@lastlanguage{\the\csname l@#1\endcsname}%
+      \language=\l@nohyphenation%
+  }{}
 }

```

@reutenauer can you remember what the initial idea was?

**Comment by dohyunkim** [comment]
polyglossia.sty has correctly implemented what the manual says. However, some gloss files (english, french, german, greek, hindi, latin, and portuguese) are overwriting `\<languagename>@language` command provided by .sty file.

For instance, `\disablehyphenation` and `\enablehyphenation` do work as expected with gloss-spanish:
```latex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{english}
\setotherlanguage{spanish}
\begin{document}
\showhyphens{hyphenation}
\disablehyphenation
\showhyphens{hyphenation}
\selectlanguage{spanish}
\showhyphens{castellano}
\enablehyphenation
\showhyphens{castellano}
\end{document}
```

**Comment by jspitz** [comment]
So we need the patch I proposed.

**Comment by dohyunkim** [comment]
The patch proposed has a problem.
```latex
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage[variant=ancientgreek]{greek}
\newfontfamily\greekfont{texgyrepagella-regular.otf}
\begin{document}
\typeout{===\the\language===}
\disablehyphenation
\selectlanguage[variant=ancientgreek]{greek}
\enablehyphenation
\typeout{===\the\language===}
\end{document}
```
It is expected that the two numbers printed are identical, but actually it is not.

**Comment by dohyunkim** [comment]
How about this patch:
```diff
diff --git a/tex/polyglossia.sty b/tex/polyglossia.sty
index e628aa5..e4972d3 100644
--- a/tex/polyglossia.sty
+++ b/tex/polyglossia.sty
@@ -268,16 +268,12 @@
     }%
   }%
   \csdef{#1@language}{%
-    \ifbool{xpg@hyphenation@disabled}{%
-      \xdef\xpg@lastlanguage{\the\csname l@#1\endcsname}%
-    }{%
       \ifluatex%
         \cs_if_eq:cNF{l@#1}{\l@nohyphenation}{%
           \xpg@set@language@luatex@iv{#1}%
         }%
       \fi%
       \language=\csname l@#1\endcsname%
-    }%
   }%
   % setup hypenmins
   \exp_args:NNe \clist_set:Nn \l_tmpa_clist
@@ -1323,6 +1319,11 @@
   \use@localhyphenmins{#1}%
   \polyglossia@lang@frenchspacing:n{#1}
   \polyglossia@lang@setdirection:n{#1}
+  % If \disablehyphenation was active, re-disable hyphenation
+  \ifbool{xpg@hyphenation@disabled}{%
+      \xdef\xpg@lastlanguage{\the\language}%
+      \language=\l@nohyphenation
+  }{}%
 }
```
The point is `\the\language` instead of `\the\csname l@#1\endcsname`.

**Comment by jspitz** [comment]
I get a mismatch also without the patch. This is because Greek has no hyphenation.
I'll investigate.

**Comment by jspitz** [comment]
I have fixed it by implementing a hyphenation setup macro (for glosses with variants) that considers
hyphenation disabling.

The only case which is not yet fixed in latin (since @wehro requested not to touch that file). @wehro, please implement this fix.

**Comment by wehro** [comment]
Thank you for the improvements. However, with the following code, I don't get the expected result.
```
\documentclass{article}
\usepackage{polyglossia}
\setmainlanguage{english}
\setotherlanguage{german}

\begin{document}
This text is hyphenated.
\showhyphens{hyphenation}
\disablehyphenation
This text is not hyphenated.
\showhyphens{hyphenation}
\selectlanguage[variant=swiss]{german}
This text should not be hyphenated, but it is.
\showhyphens{Silbentrennung}
\end{document}
```

**Comment by jspitz** [comment]
> Thank you for the improvements. However, with the following code, I don't get the expected result.
> 
> ```
> \documentclass{article}
> \usepackage{polyglossia}
> \setmainlanguage{english}
> \setotherlanguage{german}
> 
> \begin{document}
> This text is hyphenated.
> \showhyphens{hyphenation}
> \disablehyphenation
> This text is not hyphenated.
> \showhyphens{hyphenation}
> \selectlanguage[variant=swiss]{german}
> This text should not be hyphenated, but it is.
> \showhyphens{Silbentrennung}
> \end{document}
> ```

This is fixed at ae08d98
